*: add a doc and related description for the Optimizer Fix Controls (#13831)
pingcap/docs: optimizer-fix-controls.md

---
title: Optimizer Fix Controls
summary: Learn about the Optimizer Fix Controls feature and how to use `tidb_opt_fix_control` to control the TiDB optimizer in a more fine-grained way.
---

# Optimizer Fix Controls

> The TiDB optimizer has lots of details in its behavior. As TiDB evolves, these details are changing. Usually, these changes are improvements of the optimizer. But sometimes, they would possibly cause unexpected results, such as:

Reviewer: ```suggestion
The TiDB optimizer has lots of details in its behavior. As TiDB evolves, these details are changing. Usually, these changes are improvements of the optimizer. But sometimes, they might cause unexpected results, such as:
```

> - For some implementation details, some behaviors are more suitable for certain scenarios. Some changes that bring improvements for most scenarios might cause regressions for other scenarios.

Reviewer: ```suggestion
- For some implementation details, some behaviors are more suitable for certain scenarios. Changes that bring improvements for most scenarios might cause regressions for other scenarios.
```

> - For some implementation details, different behaviors are more suitable to different scenarios. Some changes that bring improvements for most scenarios may also cause regressions for a few of other scenarios.

Reviewer: ```suggestion
- For some implementation details, different behaviors are more suitable to different scenarios. Some changes that bring improvements for most scenarios might also cause regressions for a few of other scenarios.
```

- Sometimes, the relationship between changes in the behavior details and their consequences is very complicated. Even if it is an improvement in a certain behavior, it might cause execution plan regression as a whole.

> 

Reviewer: ```suggestion

## `tidb_opt_fix_control`

```

---
title: Optimizer Fix Controls
summary: Learn about the Optimizer Fix Controls and how to use tidb_opt_fix_control to control TiDB optimizer in a more fine-grained way
aliases: ['/docs/dev/optimizer-fix-controls/']
---

# Optimizer Fix Controls

The TiDB optimizer has lots of details in its behavior. As TiDB evolves, these details are changing. Usually, these changes are improvements of the optimizer. But sometimes, they would possibly cause unexpected results like execution plan regression:

- For some implementation details, different behaviors are more suitable to different scenarios.
- Sometimes, the relationship between changes in the behavior details and their consequences is very complicated. Even if it's an improvement in a certain behavior, it might cause execution plan regression as a whole.

> So we provided the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v710) system variable to control the behavior of the optimizer in a more fine-grained way.

Reviewer: ```suggestion
Starting from v7.1.0, TiDB provides the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v710) system variable to control the behavior of the optimizer in a more fine-grained way.
```

> A Fix is usually denoted by a number, which usually corresponds to a GitHub Issue, in which there will be descriptions for the technical details. For example, Fix `44262` corresponds to [Issue 44262](https://github.com/pingcap/tidb/issues/44262).

Reviewer: ```suggestion
A fix is usually denoted by a number, which usually corresponds to a GitHub Issue. There are descriptions about the technical details in the GitHub issue. For example, Fix `44262` corresponds to [Issue 44262](https://github.com/pingcap/tidb/issues/44262).
```

> This variable supports multiple Fixes, separated by commas (`,`). The format is `"<#issue1>:<value1>,<#issue2>:<value2>,...,<#issueN>:<valueN>"`, where `<#issueN>` is the Fix number. For example:

Reviewer: ```suggestion
This variable supports multiple fixes, separated by commas (`,`). The format is `"<#issue1>:<value1>,<#issue2>:<value2>,...,<#issueN>:<valueN>"`, where `<#issueN>` is the fix number. For example:
```

```sql
SET SESSION tidb_opt_fix_control = '44262:ON,44389:ON';
```

> ## Optimizer Fix Controls Reference

Reviewer: ```suggestion
## Optimizer Fix Controls reference
```

### [`44262`](https://github.com/pingcap/tidb/issues/44262) <span class="version-mark">New in v7.1.1</span>

- Default value: `OFF`
- Possible values: `ON`, `OFF`
- This variable controls whether to allow the use of [Dynamic pruning mode](/partitioned-table.md#dynamic-pruning-mode) to access the partitioned table when the [GlobalStats](/statistics.md#collect-statistics-of-partitioned-tables-in-dynamic-pruning-mode) are missing.

### [`44389`](https://github.com/pingcap/tidb/issues/44389) <span class="version-mark">New in v7.1.1</span>

- Default value: `OFF`
- Possible values: `ON`, `OFF`

> - For filters like `c = 10 and (a = 'xx' or (a = 'kk' and b = 1)`, whether to try to build more complete scan ranges for `IndexRangeScan`.

Reviewer: ```suggestion
- For filters such as `c = 10 and (a = 'xx' or (a = 'kk' and b = 1)`, this variable controls whether to try to build more comprehensive scan ranges for `IndexRangeScan`.
```

---
title: Optimizer Fix Controls
summary: Learn about the Optimizer Fix Controls feature and how to use `tidb_opt_fix_control` to control the TiDB optimizer in a more fine-grained way.
---

# Optimizer Fix Controls

> The TiDB optimizer has lots of details in its behavior. As TiDB evolves, these details are changing. Usually, these changes are improvements of the optimizer. But sometimes, they might cause unexpected results, such as:

Reviewer: ```suggestion
As the product evolves iteratively, the behavior of the TiDB optimizer changes, which in turn generates more reasonable execution plans. However, in some specific scenarios, the new behavior may lead to unintended results. For example:
```

> As the product evolves iteratively, the behavior of the TiDB optimizer changes, which in turn generates more reasonable execution plans. However, in some specific scenarios, the new behavior may lead to unintended results. For example:

Reviewer 2: ```suggestion
As the product evolves iteratively, the behavior of the TiDB optimizer changes, which in turn generates more reasonable execution plans. However, in some particular scenarios, the new behavior may lead to unexpected results. For example:
```

> - For some implementation details, some behaviors are more suitable for certain scenarios. Changes that bring improvements for some scenarios might cause regressions for others.

Reviewer: ```suggestion
- The effect of some behaviors is related to a specific scenario. Changes that bring improvements for some scenarios might cause regressions for others.
```

> - Sometimes, the relationship between changes in the behavior details and their consequences is very complicated. An improvement in a certain behavior might cause execution plan regression as a whole.

Reviewer 2: ```suggestion
- Sometimes, the relationship between changes in the behavior details and their consequences is very complicated. An improvement in a certain behavior might cause overall regression.
```

> Therefore, TiDB provides the Optimizer Fix Controls feature that allows you to control the details of TiDB optimizer behaviors by setting a series of fixes. This document describes the Optimizer Fix Controls feature and how to use them, and lists all the fixes that TiDB currently supports.

Reviewer 2: ```suggestion
Therefore, TiDB provides the Optimizer Fix Controls feature that allows you to make fine-grained control of TiDB optimizer behaviors by setting values for a group of fixes. This document describes the Optimizer Fix Controls feature and how to use them, and lists all the fixes that TiDB currently supports for Fix Controls.
```

## Introduction to `tidb_opt_fix_control`

Starting from v7.1.0, TiDB provides the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v710) system variable to control the behavior of the optimizer in a more fine-grained way.

> A fix is a control item used to adjust the behavior in the TiDB optimizer at one point. It is denoted by a number that corresponds to a GitHub Issue. The GitHub issue contains the technical details. For example, Fix `44262` corresponds to [Issue 44262](https://github.com/pingcap/tidb/issues/44262).

Reviewer 2: ```suggestion
Each fix is a control item used to adjust the behavior in the TiDB optimizer for one particular purpose. It is denoted by a number that corresponds to a GitHub Issue, that contains the technical details of the behavior change. For example, for fix `44262`, you can review what it controls in [Issue 44262](https://github.com/pingcap/tidb/issues/44262).
```

> This variable supports multiple fixes, separated by commas (`,`). The format is `"<#issue1>:<value1>,<#issue2>:<value2>,...,<#issueN>:<valueN>"`, where `<#issueN>` is the fix number. For example:

Reviewer 2: ```suggestion
Variable [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v710) accepts multiple fixes as one value, separated by commas (`,`). The format is `"<#issue1>:<value1>,<#issue2>:<value2>,...,<#issueN>:<valueN>"`, where `<#issueN>` is the fix number. For example:
```
